Docs/configuration.md
# Configuration

This document is meant to be an exhaustive reference of all TOML configuration fields available in Steward. Some of these fields are also used by the Orchestrator.

You can see all available configuration fields with their default values by running

```bash
steward print-config
```

To generate a default config file, simply redirect stdout to a file:

```bash
# You must update these values to your own configuration
steward print-config > config.toml
```

## Reference by Table

### `[cosmos]` table

Configuration related interactions with the Cosmos chain in question

#### `gas_adjustment`

Type: float

Multiplied by estimated gas fee per transaction. Currently Sommelier fee requirements are 0.0 so this can be left as default.

```
[cosmos]
gas_adjustment = 1.0
```

#### `grpc`

Type: string

Your validator node's gRPC endpoint

```
[cosmos]
grpc = "http://localhost:9090"
```

#### `key_derivation_path`

Type: string

The path used to derive accounts from your private keys. For Steward to work this must match whatever derivation path was used for the delegate and operator accounts you registered in the validator set on-chain.

```
[cosmos]
key_derivation_path = "m/44'/118'/0'/0/0"
```

#### `msg_batch_size`

Type: integer

Used by the Orchestrator. The max number of Msgs to send to Sommelier at once.

```
[cosmos]
msg_batch_size = 5
```

#### `prefix`

Type: string

The bech32 prefix to use for representing account addresses

```
[cosmos]
prefix = "somm"
```

### `[cosmos.gas_price]` table

For setting transaction fees

#### `amount`

Type: float

The amount of `denom` you are willing to pay per gas for transactions.

```
[cosmos.gas_price]
amount = 0.0
```

#### `denom`

Type: string

The denomination of the gas fee `amount`. Sommelier handles fees in `usomm`.

```
[cosmos.gas_price]
denom = "usomm"
```

### `[ethereum]` table

Config related to interaction with Ethereum

#### `blocks_to_search`

Type: integer

How many Ethereum blocks back the Orchestrator should search for the last event nonce relayed to Cosmos.

```
[ethereum]
blocks_to_search = 5000
```

#### `gas_price_multiplier`

Type: float

Multiplied by estimated gas fee per transaction. If your Ethereum transactions are failing due to insufficient gas, try increasing this value.

```
[ethereum]
gas_price_multiplier = 1.0
```

#### `key_derivation_path`

Type: string

The path used to derive Etherum addresses from your private keys.

```
[ethereum]
key_derivation_path = "m/44'/60'/0'/0/0"
```

#### `rpc`

Type: string

The node endpoint to be used for Ethereum interactions

```
[ethereum]
rpc = "http://localhost:8545"
```

### `[gravity]` table

Configuration related to the gravity bridge

#### `contract`

Type: string

The address of the gravity contract on Ethereum

```
[gravity]
contract = "0x0000000000000000000000000000000000000000"
```

#### `fees_denom`

Type: string

The denomination in which fees will be paid to relayers. This value must be a denom supported by the configured chain (whatever chain the node at `cosmos.grpc` is validating on).

```
[gravity]
fees_denom = "usomm"
```

### `[keys]` table

Keys config

#### `delegate_key`

Type: string

The name of the key in `keystore` used for delegate signing by both Steward and the Orchestrator. The key name is the name of the key file *without the file extension*.

```
[keys]
delegate_key = ""
```

#### `keystore`

Type: string

Path to the location where `steward keys` will manage keys on-disk

```
[keys]
keystore = "/tmp/keystore"
```


### [metrics] table

Config related to the Orchestrator metrics server

#### `listen_addr`

Type: string

The server endpoint for monitoring Orchestrator metrics

```
[metrics]
listen_addr = "127.0.0.1:3000"
```

### [server] table

Config related to the Steward server

#### `address`

Type: string

The IP address at which the Steward server will run

```
[server]
address = "0.0.0.0"
```

#### `client_ca_cert_path`

Type: string

The path to the trusted CA used to sign the Strategy Provider's client certificate

```
[server]
client_ca_cert_path = ""
```

#### `port`

Type: integer

The port on which the Steward server will be hosted

```
[server]
port = 5734
```

#### `server_cert_path`

Type: string

The path to your signed Steward server certificate

```
[server]
server_cert_path = ""
```

#### `server_key_path`

Type: string

The path to the PKCS8-formatted key used to generate the server certificate

```
[server]
server_key_path = ""
```

## Complete Example config.toml

This example will not work as is, you'll need to supply your own values.

```toml
keystore = "/tmp/keystore"

[cosmos]
gas_adjustment = 1.0
grpc = "https://127.0.0.1:9090"
key_derivation_path = "m/44'/118'/0'/0/0"
prefix = "somm"
msg_batch_size = 5

[cosmos.gas_price]
amount = 0.0
denom = "usomm"

[ethereum]
blocks_to_search = 5000
gas_price_multiplier = 1.0
key_derivation_path = "m/44'/60'/0'/0/0"
rpc = "http://localhost:8545"

[gravity]
contract = "0x0000000000000000000000000000000000000000"
fees_denom = "stake"

[keys]
delegate_key = ""

[metrics]
listen_addr = "127.0.0.1:3000"

[server]
address = "0.0.0.0"
client_ca_cert_path = ""
port = 5734
server_cert_path = ""
server_key_path = ""
```


docs/configuration.md
# Configuration

This document is meant to be an exhaustive reference of all TOML configuration fields available in Steward. Some of these fields are also used by the Orchestrator.

## Default configuration

You can see all available configuration fields with their default values by running

```bash
steward print-config
```

To generate a default config file, simply redirect stdout to a file:

```bash
# You must update these values to your own configuration
steward print-config > config.toml
```

## Reference

### `[cosmos]` table

Configuration related interactions with the Cosmos chain in question

#### `gas_adjustment`

Type: float

Multiplied by estimated gas fee per transaction. Currently Sommelier fee requirements are 0.0 so this can be left as default.

```
[cosmos]
gas_adjustment = 1.0
```

#### `grpc`

Type: string

Your validator node's gRPC endpoint

```
[cosmos]
grpc = "http://localhost:9090"
```

#### `key_derivation_path`

Type: string

The path used to derive accounts from your private keys. For Steward to work this must match whatever derivation path was used for the delegate and operator accounts you registered in the validator set on-chain.

```
[cosmos]
key_derivation_path = "m/44'/118'/0'/0/0"
```

#### `msg_batch_size`

Type: integer

Used by the Orchestrator. The max number of Msgs to send to Sommelier at once.

```
[cosmos]
msg_batch_size = 5
```

#### `prefix`

Type: string

The bech32 prefix to use for representing account addresses

```
[cosmos]
prefix = "somm"
```

### `[cosmos.gas_price]` table

For setting transaction fees

#### `amount`

Type: float

The amount of `denom` you are willing to pay per gas for transactions.

```
[cosmos.gas_price]
amount = 0.0
```

#### `denom`

Type: string

The denomination of the gas fee `amount`. Sommelier handles fees in `usomm`.

```
[cosmos.gas_price]
denom = "usomm"
```

### `[ethereum]` table

Config related to interaction with Ethereum

#### `blocks_to_search`

Type: integer

How many Ethereum blocks back the Orchestrator should search for the last event nonce relayed to Cosmos.

```
[ethereum]
blocks_to_search = 5000
```

#### `gas_price_multiplier`

Type: float

Multiplied by estimated gas fee per transaction. If your Ethereum transactions are failing due to insufficient gas, try increasing this value.

```
[ethereum]
gas_price_multiplier = 1.0
```

#### `key_derivation_path`

Type: string

The path used to derive Etherum addresses from your private keys.

```
[ethereum]
key_derivation_path = "m/44'/60'/0'/0/0"
```

#### `rpc`

Type: string

The node endpoint to be used for Ethereum interactions

```
[ethereum]
rpc = "http://localhost:8545"
```

### `[gravity]` table

Configuration related to the gravity bridge

#### `contract`

Type: string

The address of the gravity contract on Ethereum

```
[gravity]
contract = "0x0000000000000000000000000000000000000000"
```

#### `fees_denom`

Type: string

The denomination in which fees will be paid to relayers. This value must be a denom supported by the configured chain (whatever chain the node at `cosmos.grpc` is validating on).

```
[gravity]
fees_denom = "usomm"
```

### `[keys]` table

Keys config

#### `delegate_key`

Type: string

The name of the key in `keystore` used for delegate signing by both Steward and the Orchestrator. The key name is the name of the key file *without the file extension*.

```
[keys]
delegate_key = ""
```

#### `keystore`

Type: string

Path to the location where `steward keys` will manage keys on-disk

```
[keys]
keystore = "/tmp/keystore"
```


### [metrics] table

Config related to the Orchestrator metrics server

#### `listen_addr`

Type: string

The server endpoint for monitoring Orchestrator metrics

```
[metrics]
listen_addr = "127.0.0.1:3000"
```

### [server] table

Config related to the Steward server

#### `address`

Type: string

The IP address at which the Steward server will run

```
[server]
address = "0.0.0.0"
```

#### `client_ca_cert_path`

Type: string

The path to the trusted CA used to sign the Strategy Provider's client certificate

```
[server]
client_ca_cert_path = ""
```

#### `port`

Type: integer

The port on which the Steward server will be hosted

```
[server]
port = 5734
```

#### `server_cert_path`

Type: string

The path to your signed Steward server certificate

```
[server]
server_cert_path = ""
```

#### `server_key_path`

Type: string

The path to the PKCS8-formatted key used to generate the server certificate

```
[server]
server_key_path = ""
```

## Complete Example config.toml

This example will not work as is, you'll need to supply your own values.

```toml
keystore = "/tmp/keystore"

[cosmos]
gas_adjustment = 1.0
grpc = "https://127.0.0.1:9090"
key_derivation_path = "m/44'/118'/0'/0/0"
prefix = "somm"
msg_batch_size = 5

[cosmos.gas_price]
amount = 0.0
denom = "usomm"

[ethereum]
blocks_to_search = 5000
gas_price_multiplier = 1.0
key_derivation_path = "m/44'/60'/0'/0/0"
rpc = "http://localhost:8545"

[gravity]
contract = "0x0000000000000000000000000000000000000000"
fees_denom = "stake"

[keys]
delegate_key = ""

[metrics]
listen_addr = "127.0.0.1:3000"

[server]
address = "0.0.0.0"
client_ca_cert_path = ""
port = 5734
server_cert_path = ""
server_key_path = ""
```


steward/src/config.rs
//! Steward Config
//!
//! See instructions in `commands.rs` to specify the path to your
//! application's configuration file and/or command-line options
//! for specifying it.
use crate::prelude::APP;
use abscissa_core::Application;
use deep_space::Address;
use ethers::{prelude::H160, signers::LocalWallet as EthWallet};
use gravity_bridge::cosmos_gravity;
use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use signatory::FsKeyStore;
use std::{net::SocketAddr, path::Path, time::Duration};

lazy_static! {
    pub static ref DELEGATE_KEY: cosmos_gravity::crypto::PrivateKey = {
        let config = APP.config();
        let name = &config.keys.delegate_key;
        config.load_deep_space_key(name.clone())
    };
    pub static ref DELEGATE_ADDRESS: Address = {
        let config = APP.config();
        DELEGATE_KEY
            .to_address(&config.cosmos.prefix)
            .expect("failed to derive delegate address from key. make sure the cosmos.prefix field is set in your config.")
    };
}

/// Steward Configuration
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct StewardConfig {
    pub keystore: String,
    pub cellars: Vec<CellarConfig>,
    pub cosmos: CosmosSection,
    pub ethereum: EthereumSection,
    pub gravity: GravitySection,
    pub keys: KeysConfig,
    pub metrics: MetricsSection,
    pub server: ServerSection,
}

impl StewardConfig {
    fn load_secret_key(&self, name: String) -> k256::elliptic_curve::SecretKey<k256::Secp256k1> {
        let keystore = Path::new(&self.keystore);
        let keystore = FsKeyStore::create_or_open(keystore).expect("Could not open keystore");
        let name = name.parse().expect("Could not parse name");
        let key = keystore.load(&name).expect("Could not load key");
        key.to_pem().parse().expect("Could not parse pem")
    }

    pub fn load_clarity_key(&self, name: String) -> clarity::PrivateKey {
        let key = self.load_secret_key(name).to_bytes();
        clarity::PrivateKey::from_slice(&key).expect("Could not convert key")
    }

    pub fn load_deep_space_key(&self, name: String) -> cosmos_gravity::crypto::PrivateKey {
        let key = self.load_secret_key(name).to_bytes();
        let key = deep_space::utils::bytes_to_hex_str(&key);
        key.parse().expect("Could not parse private key")
    }

    pub fn load_ethers_wallet(&self, name: String) -> EthWallet {
        EthWallet::from(self.load_secret_key(name))
    }
}

/// Default configuration settings.
///
/// Note: if your needs are as simple as below, you can
/// use `#[derive(Default)]` on StewardConfig instead.
impl Default for StewardConfig {
    fn default() -> Self {
        Self {
            keystore: String::new(),
            cellars: vec![CellarConfig::default()],
            cosmos: CosmosSection::default(),
            ethereum: EthereumSection::default(),
            gravity: GravitySection::default(),
            keys: KeysConfig::default(),
            metrics: MetricsSection::default(),
            server: ServerSection::default(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct ServerSection {
    pub address: String,
    pub client_ca_cert_path: Option<String>,
    pub port: u16,
    pub server_cert_path: String,
    pub server_key_path: String,
}

impl Default for ServerSection {
    fn default() -> Self {
        Self {
            address: "0.0.0.0".to_string(),
            client_ca_cert_path: Some("".to_owned()),
            port: 5734,
            server_cert_path: "".to_owned(),
            server_key_path: "".to_owned(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct KeysConfig {
    pub delegate_key: String,
}

impl Default for KeysConfig {
    fn default() -> Self {
        Self {
            delegate_key: "".to_owned(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct CellarConfig {
    pub pair_id: ethers::types::U256,
    pub name: String,
    pub token_id: ethers::types::U256,
    pub cellar_address: ethers::types::H160,
    pub pool_address: ethers::types::H160,
    pub weight_factor: u32,
    pub max_gas_price_gwei: u32,
    pub token_0: TokenInfo,
    pub token_1: TokenInfo,
    pub duration: Duration,
}

impl Default for CellarConfig {
    fn default() -> Self {
        CellarConfig {
            pair_id: ethers::types::U256::zero(),
            name: String::default(),
            token_id: ethers::types::U256::zero(),
            cellar_address: ethers::types::H160::zero(),
            pool_address: ethers::types::H160::zero(),
            weight_factor: 100,
            max_gas_price_gwei: 100,
            duration: Duration::from_secs(60),
            token_0: TokenInfo::default(),
            token_1: TokenInfo::default(),
        }
    }
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct TokenInfo {
    pub decimals: u8,
    pub symbol: String,
    pub address: H160,
}

impl Default for TokenInfo {
    fn default() -> Self {
        TokenInfo {
            decimals: 18,
            symbol: "NA".to_string(),
            address: H160::zero(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct EthereumSection {
    pub blocks_to_search: u64,
    pub gas_price_multiplier: f32,
    pub key_derivation_path: String,
    pub rpc: String,
}

impl Default for EthereumSection {
    fn default() -> Self {
        Self {
            blocks_to_search: 5000,
            gas_price_multiplier: 1.0f32,
            key_derivation_path: "m/44'/60'/0'/0/0".to_owned(),
            rpc: "http://localhost:8545".to_owned(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct GravitySection {
    pub contract: String,
    pub fees_denom: String,
}

impl Default for GravitySection {
    fn default() -> Self {
        Self {
            contract: "0x0000000000000000000000000000000000000000".to_owned(),
            fees_denom: "usomm".to_owned(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct CosmosSection {
    pub gas_adjustment: f64,
    pub grpc: String,
    pub key_derivation_path: String,
    pub prefix: String,
    pub msg_batch_size: u32,
    pub gas_price: GasPrice,
}

impl Default for CosmosSection {
    fn default() -> Self {
        Self {
            gas_adjustment: 1.0f64,
            grpc: "https://127.0.0.1:9090".to_owned(),
            key_derivation_path: "m/44'/118'/0'/0/0".to_owned(),
            prefix: "somm".to_owned(),
            msg_batch_size: 5,
            gas_price: GasPrice::default(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct GasPrice {
    pub amount: f64,
    pub denom: String,
}

impl Default for GasPrice {
    fn default() -> Self {
        Self {
            amount: 0.0,
            denom: "usomm".to_owned(),
        }
    }
}

impl GasPrice {
    pub fn as_tuple(&self) -> (f64, String) {
        (self.amount, self.denom.to_owned())
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default, deny_unknown_fields)]
pub struct MetricsSection {
    pub listen_addr: SocketAddr,
}

impl Default for MetricsSection {
    fn default() -> Self {
        Self {
            listen_addr: "127.0.0.1:3000".parse().unwrap(),
        }
    }
}
